ksql-api/src/main/java/io/confluent/ksql/api/plugin/KsqlServerEndpoints.java

```diff
@@ -168,10 +176,43 @@ public String getName() {
   }
 
   @Override
-  public InsertsSubscriber createInsertsSubscriber(final String target,
+  public InsertsStreamSubscriber createInsertsSubscriber(final String target,
       final JsonObject properties,
-      final Subscriber<JsonObject> acksSubscriber) {
-    return null;
+      final Subscriber<InsertResult> acksSubscriber, final Context context,
+      final WorkerExecutor workerExecutor) {
+    Utils.checkIsWorker();
+    final ServiceContext serviceContext = createServiceContext(new DummyPrincipal());
+    final DataSource dataSource = getDataSource(ksqlConfig, ksqlEngine.getMetaStore(),
+        SourceName.of(target));
+    if (dataSource.getDataSourceType() == DataSourceType.KTABLE) {
+      throw new KsqlException("Cannot insert into a table");
+    }
+    return InsertsSubscriber.createInsertsSubscriber(serviceContext, properties, dataSource,
+        ksqlConfig, context, acksSubscriber, workerExecutor);
+  }
+
+  private DataSource getDataSource(
+      final KsqlConfig ksqlConfig,
+      final MetaStore metaStore,
+      final SourceName sourceName
+  ) {
+    final DataSource dataSource = metaStore.getSource(sourceName);
+    if (dataSource == null) {
+      throw new KsqlException("Cannot insert values into an unknown stream: "
+          + sourceName);
+    }
+
+    if (dataSource.getKsqlTopic().getKeyFormat().isWindowed()) {
+      throw new KsqlException("Cannot insert values into windowed stream");
+    }
+
+    final ReservedInternalTopics internalTopics = new ReservedInternalTopics(ksqlConfig);
```

Comment:
`internalTopics` can probably be a field.  You could even inject it so that it's easier to unit test this class without coupling to `ReservedInternalTopics` functionality.

Reply:
ack

Reply:
> Interactive users obviously still want to interact using SQL, so we're going to need the SQL `INSERT VALUES` functionality at some point. Do you see this being handled through a standard SQL processing endpoint? 

AIUI, currently INSERT VALUES is executed via the /ksql endpoint (although doesn't seem to be documented in the REST API docs https://docs.confluent.io/current/ksql/docs/developer-guide/api.html so I didn't have a task created for this).

Current plan is to just support it in the new API as is.

> One thing the UI team struggle with is how different statements need to go to different endpoints in the current API, e.g. queries and print topics to a websocket APi. The challenge is that the UI doesn't know what statement is which without some form of parsing - which couples the UI to the SQL syntax. If would be good if the new API could think of a way to solve this issue and be more UI friendly. (I'd previously though of adding an endpoint that would just parse the text a user entered in the UI but not execute it. This could then be returned to the UI to then direct the different bits at the different endpoints - while this would have helped with the old API, these feels rather convoluted for a new API).

I suppose if the user just wants to execute a command and display the result as text on a UI somewhere then we can create an endpoint that allows any command to be executed and returned that way. This means they don't have to figure out what endpoint to call.

If they want a structured response (e.g. a list of Json arrays) so they can be easily parsed then they're going to have to have different code for parsing the response depending on what command they execute, so they're going to have to know what the command is at the client anyway, so calling a different endpoint for those is no extra pain.

We could create such a generic endpoint that can be accessed over http or websockets.

> 
> There are two things I'd really like to see added to this API soon:
> 
> 1. Authentication and authorisation - we can't turn the thing on until we have it and its better to do it from the ground up rather than to bolt it on after.

Ack. @vcrfxia  is working on the basic auth integration. 

> 2. A JSON based test framework, i.e. along the lines of `QueryTranslationTest`.  Something that can just accept the JSON requests and assert the JSON responses.

Seems like a good idea.